File: x1/busnumbers/busnumbers.hs
import Data.List(sort, intercalate)

-- Run: 
--      ghc busnumbers.hs
--      ./busnumbers

-- Function that converts the input string into a list of integers instead
-- words: splits string into list of words
-- read: casts a string to another type (in this case integer)
-- map: casts functions read and words on every char in input string
convertToArraylist :: String -> [Int]
convertToArraylist listOfBuslines = map read (words listOfBuslines)

-- Function that groups every consecutive number into lists
--      EXAMPLE: input = [1, 2, 3, 5, 4], output = [[1, 2, 3], [4, 5]]
-- Base case: return empty list for empty list input
-- Recursive case: sort input list in ascending order, call local func 'go' with empty accumulator
groupConsecutive :: [Int] -> [[Int]]
groupConsecutive [] = []
groupConsecutive xs = go (sort xs) []
    where
        go :: [Int] -> [Int] -> [[Int]] 
        go [] acc = [reverse acc] -- if no more nums left in xs, reverse acc
        go [x] acc = [reverse (x:acc)] -- if one element remains, add to acc and reverse 
        go (x:y:ys) acc -- if there are two or more elements left
            | y == x+1 = go (y:ys) (x:acc) -- if x and y are consecutive, 
            | otherwise = reverse (x:acc) : go (y:ys) [] -- if not consecutive, add x and reverse, start new group with y

-- function that takes the grouped list of ints and creates a string with correct format
createString :: [[Int]] -> String
createString sortedBuslines = unwords (map go sortedBuslines)
    where
        go :: [Int] -> String
        go innerlist
            | length innerlist > 2 = show (head innerlist) ++ "-" ++ show (last innerlist) 
            | otherwise = unwords (map show innerlist)

-- mainfunc to read from terminal
main :: IO()
main = do
    -- amount of buslines
    buslines <- getLine

    -- list of buslines
    buslist <- getLine
    let getList = groupConsecutive (convertToArraylist buslist)
    let formattedList = createString getList

    putStrLn formattedList

File: x1/busnumbers/busnumbers.java
// Run: 
//      javac busnumbers.java
//      java busnumbers.java

package busnumbers;

import java.util.*;

public class busnumbers {
    static int numberOfBuses = 0;
    static List<Integer> listOfBuslines = new ArrayList<Integer>();
    static List<String> formattedListOfBuslines = new ArrayList<String>();
    
    // Function that takes string of numbers and returns a list of ints 
    public static List<Integer> convertToArraylist (String numbers) {
        // store each string split by single spaces in numberStrings array
        String[] numberStrings = numbers.split("\s"); // Separate with regex for single spaces
        
        // Loop the list of strings and convert each element to int and add to global int list
        for (int i = 0; i < numberStrings.length; i++) {
            listOfBuslines.add(Integer.parseInt(numberStrings[i]));
        }
        // Return the int list
        return listOfBuslines;
    }

    public static List<String> findConsecutive(List<Integer> inputNumbers) {
        List<String> formattedNumbers = new ArrayList<String>(); // formatted numbers in a stringlist
        Collections.sort(inputNumbers); // sort input numbers for correct output order

        // loop through the input numbers
        for (int i = 0; i < inputNumbers.size(); i++) {
            int currentNum = inputNumbers.get(i); // Retrieve current number
            List<Integer> consecNums = new ArrayList<Integer>(); // List to store consec numbers
            consecNums.add(currentNum); // add current to consec list

            // while next num is consecutive, keep adding to consecNums
            while (i+1 < inputNumbers.size() && inputNumbers.get(i+1) == inputNumbers.get(i)+1) {
                i++;
                consecNums.add(inputNumbers.get(i));
            }
            // If consecNums has more than 2 elems, its worth it to save space
            if (consecNums.size() > 2) {
                String firstNum = String.valueOf(consecNums.get(0)); // stringify first num
                String lastNum = String.valueOf(consecNums.get(consecNums.size()-1)); // stringify last num
                String formatted = firstNum + "-" + lastNum; // concat to correct format

                formattedNumbers.add(formatted); // add formatted string to list of formatted nums
            }
            // otherwise, add string as single elems
            else {
                // for each elem in consecNums (1 or 2 elems)
                for (int num : consecNums) {
                    String current = String.valueOf(num); // stringify Int
                    formattedNumbers.add(current); // add to formatted nums
                }
            }
        }
        return formattedNumbers; // return list of formatted nums
    }
    
    // func to print out array in one line
    public static void printBuslines(List<String> inputStrings) {
        for (int i = 0; i < inputStrings.size(); i++) {
            String currentString = inputStrings.get(i);
            System.out.print(currentString + " "); // print in one line and spece between
        }
        System.out.println(); // create newline after printing out all buslines
    }
    
    // Main method for input in terminal
    public static void main(String[] args) {
        // Scanner to read lines from terminal input
        Scanner scanner = new Scanner(System.in);

        // store terminal input in variable buslines (list length)
        String buslines = scanner.nextLine();
        numberOfBuses = Integer.parseInt(buslines); // set global variable from 0 to input num

        // store terminal input in variable input2 (list of buslines)
        String buslist = scanner.nextLine(); 
        listOfBuslines = convertToArraylist(buslist); // change value of global list  
        formattedListOfBuslines = findConsecutive(listOfBuslines); // reformat listOfBuslines with func

        // print in correct way
        printBuslines(formattedListOfBuslines);

        // close scanner
        scanner.close();
    }
}